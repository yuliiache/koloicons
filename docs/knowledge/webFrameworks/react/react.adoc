== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







