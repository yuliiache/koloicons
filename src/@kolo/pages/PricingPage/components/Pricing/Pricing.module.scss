@import "@kolo/design";

.pricing {
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  align-items: stretch;
  box-sizing: border-box;
  width: 100%;
  max-width: 51.2rem;
  margin: 0 auto;
  padding: 2.4rem 0.8rem 0.8rem;
}

.pricingCards {
  display: flex;
  margin: 0 -0.8rem;
}

@media screen and (max-width: $mobileWidth) {
  .pricing {
    padding: 0.8rem 0;
  }

  .pricingCards {
    flex-wrap: wrap;
    margin: 0 -0.4rem;
  }
}
