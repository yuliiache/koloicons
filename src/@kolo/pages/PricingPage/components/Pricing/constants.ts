export enum Currency {
  USD = 'USD',
  EUR = 'EUR',
  // Add more as needed
}

export enum SubscriptionPlan {
  PERSONAL = 'Personal',
  BUSINESS = 'Business',
  EXTENDED = 'Extended',
}
