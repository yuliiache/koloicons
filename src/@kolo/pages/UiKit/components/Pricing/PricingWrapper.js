import Pricing from '@kolo/pages/PricingPage/components/Pricing/Pricing';

import {Section} from '../Section/Section';

const PricingWrapper = () => {
  return (
    <Section title="Pricing">
      <Pricing />
    </Section>
  );
};

export default PricingWrapper;
