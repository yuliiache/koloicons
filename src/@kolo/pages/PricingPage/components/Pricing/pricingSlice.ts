import {BillingInterval} from '@kolo/constants/constants';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';

import {PricingData, PricingState} from './types/pricingStateTypes';

export const PRICING_KEY = 'pricing';

const initialState: PricingState = {
  isLoading: false,
  pricingPlans: [],
};

const pricingSlice = createSlice({
  name: PRICING_KEY,
  initialState,
  reducers: {
    getPricingStart: (state: PricingState, action: PayloadAction<BillingInterval>): PricingState => ({
      ...state,
      isLoading: true,
    }),
    getPricingSuccess: (state: PricingState, action: PayloadAction<PricingData[]>): PricingState => ({
      ...state,
      pricingPlans: action.payload,
      isLoading: false,
    }),
    getPricingFail: (state: PricingState): PricingState => ({
      ...state,
      pricingPlans: initialState.pricingPlans,
      isLoading: false,
    }),
  },
});

export const {getPricingStart, getPricingSuccess, getPricingFail} = pricingSlice.actions;

export default pricingSlice.reducer;
