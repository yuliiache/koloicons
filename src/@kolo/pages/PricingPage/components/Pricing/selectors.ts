import {RootState} from 'app/store';
import {createSelector} from 'reselect';

import {PRICING_KEY} from './pricingSlice';
import {PricingData, PricingState} from './types/pricingStateTypes';

const getPricing = (state: RootState): PricingState => state[PRICING_KEY];

export const pricingSelector = createSelector(
  getPricing,
  (pricing: PricingState): PricingData[] => pricing.pricingPlans
);
