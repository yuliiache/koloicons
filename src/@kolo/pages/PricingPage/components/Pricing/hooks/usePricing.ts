import {BillingInterval} from '@kolo/constants/constants';
import {useAppDispatch, useAppSelector} from 'app/hooks';
import {useEffect} from 'react';

import {getPricingStart} from '../pricingSlice';
import {pricingSelector} from '../selectors';
import {PricingData} from '../types/pricingStateTypes';

interface UsePricingReturn {
  pricingPlans: PricingData[];
  handleBillingPeriodChange: (interval: BillingInterval) => void;
}

const usePricing = (): UsePricingReturn => {
  const dispatch = useAppDispatch();
  const pricingPlans: PricingData[] = useAppSelector(pricingSelector);

  useEffect(() => {
    dispatch(getPricingStart(BillingInterval.YEAR));
  }, []);

  const handleBillingPeriodChange = (interval: BillingInterval): void => {
    dispatch(getPricingStart(interval));
  };

  return {pricingPlans, handleBillingPeriodChange};
};

export default usePricing;
