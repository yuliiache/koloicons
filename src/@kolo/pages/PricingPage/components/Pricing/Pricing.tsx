import React from 'react';

import BillingPeriodSwitcher from './components/BillingPeriodSwitcher/BillingPeriodSwitcher';
import PricingCardGallery from './components/PricingCardGallery/PricingCardGallery';
import usePricing from './hooks/usePricing';
import styles from './Pricing.module.scss';

const {pricing} = styles;

const Pricing: React.FC = () => {
  const {pricingPlans, handleBillingPeriodChange} = usePricing();

  return (
    <div className={pricing}>
      <BillingPeriodSwitcher onChange={handleBillingPeriodChange} />
      <PricingCardGallery pricingPlans={pricingPlans} />
    </div>
  );
};

export default Pricing;
