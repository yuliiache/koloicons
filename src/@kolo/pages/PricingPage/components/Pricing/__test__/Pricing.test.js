import {render, screen} from '@testing-library/react';

import usePricing from '../hooks/usePricing';
import Pricing from '../Pricing';

jest.mock('../hooks/usePricing');

jest.mock('@kolo/pages/PricingPage/components/Pricing/components/BillingPeriodSwitcher/BillingPeriodSwitcher', () => {
  const MockBillingPeriodSwitcher = () => <div data-testid="BillingPeriodSwitcherComponent" />;
  return MockBillingPeriodSwitcher;
});
jest.mock('@kolo/pages/PricingPage/components/Pricing/components/PricingCardGallery/PricingCardGallery', () => {
  const MockPricingCardGallery = () => <div data-testid="PricingCardGalleryComponent" />;
  return MockPricingCardGallery;
});

describe('Pricing Component', () => {
  const mockPricing = [
    {
      'subscriptionPlan': 'Personal',
      'paymentId': 635769,
      'features': ['SVG format', 'PNG - all sizes', '2 styles', 'Unlimited downloads', 'Updates'],
      'options': {
        'price': 2.5,
        'currency': 'USD',
        'period': 'per month, $30 billed yearly',
      },
      'license': {
        'allowed': ['End products not for sale', 'One personal social media account with non-commercial activities'],
        'disallowed': [
          'Business social media accounts',
          'Physical or digital paid advertisements',
          'Native apps, web apps, or games',
          'Resell, share, sublicense, transfer or otherwise redistribute the icons (even for free or within a more complex downloadable file)',
          'Maximum of 100 per product',
          'Trademark and Logo',
        ],
      },
    },
  ];
  const usePricingHook = () =>
    usePricing.mockReturnValue({pricingPlans: mockPricing, handleBillingPeriodChange: jest.fn()});

  it('renders without crashing with data', () => {
    usePricingHook();
    render(<Pricing />);

    expect(screen.getByTestId('BillingPeriodSwitcherComponent')).toBeInTheDocument();
    expect(screen.getByTestId('PricingCardGalleryComponent')).toBeInTheDocument();
  });

  it('should render snapshot successfully', () => {
    usePricingHook();
    const {asFragment} = render(<Pricing />);

    expect(asFragment()).toMatchSnapshot();
  });
});
