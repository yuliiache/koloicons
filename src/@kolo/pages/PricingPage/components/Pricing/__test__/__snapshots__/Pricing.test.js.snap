// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pricing Component should render snapshot successfully 1`] = `
<DocumentFragment>
  <div
    class="pricing"
  >
    <div
      data-testid="BillingPeriodSwitcherComponent"
    />
    <div
      data-testid="PricingCardGalleryComponent"
    />
  </div>
</DocumentFragment>
`;
